File: RecursiveDescentParser/project2.java
/*
 *  Toufic Hamadeh
 *  UCID 31234305
 *  Java version
 */

import java.io.*;
import java.util.*;

class project2 {

    public static void main(String[] args) {
        try {
            FileInputStream fstream = new FileInputStream("input.txt");
            DataInputStream in = new DataInputStream(fstream);
            BufferedReader br = new BufferedReader(new InputStreamReader(in));
            
            while ((s = br.readLine()) != null) {
                if (A() && i == s.length()) {
                    System.out.println("The string \"" + s + "\" is in the language.");
                    i = 0;
                }
                else {
                    System.out.println("The string \"" + s + "\" is not in the language.");
                    i = 0;
                }
            }
        in.close();
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
    
    public static boolean A() {
        if (I()) {
            if (i < s.length() && s.charAt(i) == '=') {
                ++i;
                if (E()) {
                    return true;
                }
            }
        }
        if (E()) {
            return true;
        }
    return false;
    }
  
    public static boolean E () {
        if (T()) {
            if (i < s.length() && s.charAt(i) == '+' || i < s.length() && s.charAt(i) == '-') {
                ++i;
                if (E()) {
                    return true;
                }
            }
            return true;
        }
    return false;
    }
    
    public static boolean T() {
        if (F()) {
            if (i < s.length() && s.charAt(i) == '/' || i < s.length() && s.charAt(i) == '*') {
                ++i;
                if (T()) {
                    return true;
                }
            }
            return true;
        }
    return false;
    }
    
    public static boolean F() {
        if (P()) {
            if (i < s.length() && s.charAt(i) == '^') {
                ++i;
                if (F()) {
                    return true;
                }
            }
            return true;
        }
    return false;
    }
    
    public static boolean P() {
        if (I()) {
            return true;
        }
        if (L()) {
            return true;
        }
        if (U()) {
            if(I() || L()) {
               return true;
            }
        }
        if (i < s.length() && s.charAt(i) == '(') {
            ++i;
            if (A()) {
                if (i < s.length() && s.charAt(i) == ')') {
                    ++i;
                    return true;
                }
            }
        }
    return false;
    }
    
    public static boolean U() {
        if (i < s.length() && s.charAt(i) == '+'  || i < s.length() && s.charAt(i) == '-' || i < s.length() && s.charAt(i) == '!' ) {
        ++i;
        return true;
        }
    return false;
    }
    
    public static boolean I() {
        if (C()){
            if (I()) {
                return true;
            }
        return true;
        }
    return false;
    }
    
    public static boolean C() {
        if (i < s.length() && s.charAt(i) >= 'a' && s.charAt(i) <= 'z') {
            ++i;
            return true;
        }
        return false;
    }
    
    public static boolean L() {
        if (D()){
            if (L()) {
                return true;
            }
        return true;
        }
    return false;
    }
    
    public static boolean D() {
        if (i < s.length() && '0' <= s.charAt(i) && s.charAt(i) <= '9') {
            ++i;
            return true;
        }

        return false;
    }
    private static String s;
    private static int i;
}

File: C++ Projects/RecursiveDescentParser/project2.cpp
/*
 *  Toufic Hamadeh
 *  UCID 31234305
 *  C++ version
 */

#include <iostream>
#include <string>
#include <fstream>

using namespace std;

bool A(void);
bool E(void);
bool T(void);
bool F(void);
bool P(void);
bool U(void);
bool I(void);
bool C(void);
bool L(void);
bool D(void);

string s;
string z;

int main(int argc, char *argv[]) {
    ifstream in;

    in.open("input.txt" );

    if(in.good())
    {
        while (!in.eof()) {

        getline (in,s);
        z = s;

        if (A() && s == "") {
            cout << "The string \"" << z << "\" is in the language" << endl;
        }
        else {
            cout << "The string \"" << z << "\" is not in the language" << endl;
        }
    }
    in.close();
    }
    else {
        cout << "File not found." << endl;
    }

    return 0;
}

bool A(void) {
    if (I()) {
        if (s[0] == '=') {
            s = s.substr(1);
            if (E()) {
                return true;
            }
        }
    }
    else if (E()) {
        return true;
    }
    else {
        return false;
    }
}

bool E(void) {
    if (T()) {
        if (s[0] == '+' || s[0] == '-') {
            s = s.substr(1);
            if (E()) {
                return true;
            }
        }
        return true;
    }
    return false;
}

bool T(void) {
    if (F()) {
        if (s[0] == '/' || s[0] == '*') {
            s = s.substr(1);
            if (T()) {
                return true;
            }
        }
        return true;
    }
    return false;
}

bool F(void) {
    if (P()) {
        if (s[0] == '^') {
            s = s.substr(1);
            if (F()) {
                return true;
            }
        }
        return true;
    }
    return false;
}

bool P(void) {
    if (I()) {
        return true;
    }
    if (L()) {
        return true;
    }
    if (U()) {
        if(I() || L()) {
            return true;
        }
    }
    if (s[0] == '(') {
        s = s.substr(1);
        if (A()) {
            if (s[0] == ')') {
                s = s.substr(1);
                return true;
            }
        }
    }
    return false;
}

bool U(void) {
    if (s[0]=='+'  || s[0] == '-' || s[0] == '!' ) {
        s = s.substr(1);
        return true;
    }
    return false;
}

bool I(void) {
    if (C()) {
        if (I()) {
            return true;
        }
    return true;
    }
    return false;
}

bool C(void) {
    if ('a' <= s[0] && s[0] <= 'z') {
        s = s.substr(1);
        return true;
    }
    return false;
}

bool L(void) {
    if (D()) {
        if (L()) {
            return true;
        }
        return true;
    }
    return false;
}

bool D(void) {
    if ('0' <= s[0] && s[0] <= '9') {
        s = s.substr(1);
        return true;
    }
    return false;
}

